### DISCUSSION TITLE
App is started and terminated imediately

### DISCUSSION THREAD

**Discussion by vdrasutis** [Tyrrrz/CliWrap#204] [General]
Hello.
We have command line app (using Nancy lib) which starts and then in console print "Click ENTER to close app".
Using CliWrap we starting this app by passing port argument only. App starts and stops immediately - looks like CliWrap passes, adds that "ENTER"  key press and app closes.

How I could prevent this - that app would keep running?

**Comment by Tyrrrz** [comment]
CliWrap doesn't add any key presses. What's more likely is that the target app doesn't account for the possibility that the input stream may be redirected. You can try to work around it by piping a never-resolving input:

```csharp
var input = PipeSource.Create(async (_, cancellationToken) =>
	await Task.Delay(-1, cancellationToken)
);

var cmd = input | Cli.Wrap(...);
```

**Comment by vdrasutis** [comment]
Ether I do not know how to use it of it is not helping :(.
Sample how I use CliWrapper in our test to check service health...
```
public void SentRequestIsSuccessful(string description, string webSiteUrl, bool isSuccessful, ResponseStatus responseStatus, HttpStatusCode statusCode)
        {
            var webProcess = Cli.Wrap(@"D:\MyApp.exe")
                .WithWorkingDirectory(@"D:\")
                .ExecuteAsync();

            Sleep(3000);

            using (var client = new RestClient(webSiteUrl))
            {
                var request = new RestRequest("/", Method.Get);
                request.AddHeader("Accept", "*/*");
                var response = client.ExecuteGet(request);
                var errMsg = $"{description} test failed. Status: {response.ResponseStatus}. Make sure service '{webSiteUrl}' is running";
                Assert.That(response.IsSuccessful, Is.EqualTo(isSuccessful), errMsg);
                Assert.That(response.ResponseStatus, Is.EqualTo(responseStatus), errMsg);
                Assert.That(response.StatusCode, Is.EqualTo(statusCode), errMsg);
            }

            KillProcessAndChildren(webProcess.ProcessId);
        }
```

**Reply by Tyrrrz** [reply]
Do you own `MyApp.exe`? If that's the case, I'd recommend rewriting the cancellation mechanism so that it reacts to the interrupt signal (Ctrl+C) instead of Enter. That's a very standard practice in the CLI world, and CliWrap supports it via [graceful cancellation](https://github.com/Tyrrrz/CliWrap#timeout-and-cancellation). That way the app wouldn't have to rely on the console window being present (which is required to detect key presses) or standard input (which is required for processing user input).

**Comment by vdrasutis** [comment]
ok, I think found the issue.
Some details. I have FlaUI project targeting .NET 6, but I test/run applications which is targeting .NET Framework 4.7.2.

I tried to run my app using "Process" class. Like this:
```
var processStartInfo = new ProcessStartInfo
            {
                FileName = @"C:\AppDir\MyAppexe",
                Arguments = "8080",
                WorkingDirectory = @"C:\AppDir",
                UseShellExecute = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true
            };

            var process = new Process
            {
                StartInfo = processStartInfo,
                EnableRaisingEvents = true
            };
```

By checking arguments, spotted that   "UseShellExecute" is set by default to "true"  for .NET Framework and for .NET core (which also covers .NET 6 most likely - which my projects targets...). 
So by setting "UseShellExecute" to "true" - my app starts, test pass, if set to false - fails.
![image](https://user-images.githubusercontent.com/3475668/236126113-29b35aa7-ddab-4133-b4fd-a6560d5923ed.png)

**Updated:** my bad - looks like it still does not works. Note I ether refactored application to use cancelation token  (Ctrl+C) for terminate process. So it might be related to Nancy host :/
